### ISSUE TITLE
Bundling typescript files causes problems with extensions and module names

### ISSUE THREAD

**Issue by CoskunSunali** [systemjs/systemjs#1824]
So, I spent approximately two days trying various variations of meta, map, paths, packages, etc. but no luck so far.

I created two sample projects, one with `plugin-babel` and other with `plugin-typescript`.

Babel sample: https://github.com/CoskunSunali/systemjs-babel
Typescript sample: https://github.com/CoskunSunali/systemjs-typescript

Both can be run with `npm start` right after executing `npm i`.

Our use case is that we develop in typescript and we transpile the scripts into javascript for production. We prefer typescript over babel since we prefer type safety. However, as mentioned above, the same issue occurs when we use babel as well.

**TL;DR**

Entry files are `.ts` files, output files are `.js` files. `System.register` calls are created as `"module2/y.ts"` having their module names whereas the file is not a `.ts` file anymore but `.js`.

**Long story**

Consider the following typescript files:

```ts
// x.ts
export default 123;
```

```ts
// y.ts
import x from 'module2/x';

console.log(x);
```

The output generated by babel is:

```js
// x.js (not a .ts file anymore)
System.register("module2/x.ts", [], function (_export, _context) {
  "use strict";

  return {
    setters: [],
    execute: function () {
      _export("default", 123);
    }
  };
});
```

```js
// y.js (not a .ts file anymore)
System.register('module2/y.ts', ['module2/x'], function (_export, _context) {
  "use strict";

  var x;
  return {
    setters: [function (_module2X) {
      x = _module2X.default;
    }],
    execute: function () {

      console.log(x);
    }
  };
});
```

The output generated by typescript is:

```js
// x.js (not a .ts file anymore)
System.register("module2/x.ts", [], function (exports_1, context_1) {
    "use strict";

    var __moduleName = context_1 && context_1.id;
    return {
        setters: [],
        execute: function () {
            exports_1("default", 123);
        }
    };
});
```

```js
// y.js (not a .ts file anymore)
System.register("module2/y.ts", ["module2/x"], function (exports_1, context_1) {
    "use strict";

    var __moduleName = context_1 && context_1.id;
    var x_1;
    return {
        setters: [function (x_1_1) {
            x_1 = x_1_1;
        }],
        execute: function () {
            console.log(x_1.default);
        }
    };
});
```

The first problem is that, both babel and typescript register the dependencies without extension. E.g.: `module2/x` instead of `module2/x.js` or `module2/x.ts`.

The second problem is that, both babel and typescript register the modules with extension. E.g.: `module2/x.ts` and `module2/y.ts`. However, this is a production build and we don't use typescript for production. That is why we transpile.

So, there is extension somewhere, no extension somewhere else. Is this OK?

Finally, since the output of both babel and typescript are `.js` files, is it OK that the modules get registered with `.ts` extension?

How may I hook into the `normalize` (or is it called something else?) method to override the names being used for module registration? I realize there is no easy way but how do people out there solve this very basic issue I am having?

Thanks for your replies in advance.

**Comment by CoskunSunali** [comment]
In addition to what is said above, I also tried hooking into the `resolve` method but still no success, the method never gets invoked (neither `resolveDynamic`)

```ts
    let resolve = System[System.constructor.resolve];
    System[System.constructor.resolve] = function (key, parentKey) {
        console.log('XXXXXXXXXX');
        
        return resolve.call(this, key, parentKey);
    };
```

**Comment by CoskunSunali** [comment]
Seems I finally figured it after days of debugging through the code.

I hooked into the `compile` method of `systemjs-builder/compilers/esm` module.

I am not sure if this is the right approach though. Should not it be easier for systemjs-builder to figure out the output file's extension is changing, given that it knows the "outFile"?
